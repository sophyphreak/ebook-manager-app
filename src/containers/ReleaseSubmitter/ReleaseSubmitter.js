import React, { Component } from 'react';
import { isValidPrice } from 'dao-of-validation';

import nonFictionOrFictionOptions from '../../views/FormElements/options/nonFictionOrFictionOptions';
import genreOptions from '../../views/FormElements/options/genreOptions';
import promoTypeOptions from '../../views/FormElements/options/promoTypeOptions';
import releasePage1Validation from './releasePage1Validation/releasePage1Validation';
import releasePage2Validation from './releasePage2Validation/releasePage2Validation';
import releasePage3Validation from './releasePage3Validation/releasePage3Validation';
import sendReleaseToNodemailer from './sendReleaseToNodemailer/sendReleaseToNodemailer';
import ReleaseComponent from '../../views/ReleaseComponent/ReleaseComponent';

export default class ReleaseSubmitter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentPage: 'ReleasePage1',
      error: {
        message: '',
        amazonUrl: '',
        nonFictionOrFiction: '',
        genre: '',
        email: '',
        price: '',
        promoType: '',
        website: '',
        keywords: '',
        releaseText: ''
      },

      // ReleasePage1
      amazonUrl: '',
      nonFictionOrFiction: nonFictionOrFictionOptions[0],
      genre: genreOptions[0],
      subGenre: '',
      email: '',

      // ReleasePage2
      price: '',
      promoType: promoTypeOptions[0],
      startDate: null,
      endDate: null,
      calendarFocus: null,

      // ReleasePage3
      website: '',
      keywords: '',
      releaseText: ''
    };

    // ReleasePage1
    this.onAmazonUrlChange = this.onAmazonUrlChange.bind(this);
    this.onNonFictionOrFictionChange = this.onNonFictionOrFictionChange.bind(
      this
    );
    this.onGenreChange = this.onGenreChange.bind(this);
    this.onSubGenreChange = this.onSubGenreChange.bind(this);
    this.onEmailChange = this.onEmailChange.bind(this);
    this.onSubmitReleasePage1 = this.onSubmitReleasePage1.bind(this);

    // ReleasePage2
    this.onPriceChange = this.onPriceChange.bind(this);
    this.onPromoTypeChange = this.onPromoTypeChange.bind(this);
    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onSubmitReleasePage2 = this.onSubmitReleasePage2.bind(this);

    // ReleasePage3
    this.onWebsiteChange = this.onWebsiteChange.bind(this);
    this.onKeywordsChange = this.onKeywordsChange.bind(this);
    this.onReleaseTextChange = this.onReleaseTextChange.bind(this);
    this.onSubmitReleasePage3 = this.onSubmitReleasePage3.bind(this);

    this.onBack = this.onBack.bind(this);
  }

  // ReleasePage1

  onAmazonUrlChange(e) {
    const amazonUrl = e.target.value;
    this.setState(() => ({ amazonUrl }));
  }

  onNonFictionOrFictionChange(e) {
    const nonFictionOrFiction = e.target.value;
    this.setState(() => ({ nonFictionOrFiction }));
  }

  onGenreChange(e) {
    const genre = e.target.value;
    this.setState(() => ({ genre }));
  }

  onSubGenreChange(e) {
    const subGenre = e.target.value;
    this.setState(() => ({ subGenre }));
  }

  onEmailChange(e) {
    const email = e.target.value;
    this.setState(() => ({ email }));
  }

  onSubmitReleasePage1(e) {
    e.preventDefault();
    const { error, errorsExist } = releasePage1Validation(this.state);
    this.setState(() => ({ error }));
    if (errorsExist) {
      return;
    }
    console.log(this.state);
    const currentPage = 'ReleasePage2';
    this.setState(() => ({ currentPage }));
  }

  // ReleasePage2

  onPriceChange(e) {
    const price = e.target.value;
    if (isValidPrice(price)) {
      this.setState(() => ({ price }));
    }
  }

  onPromoTypeChange(e) {
    const promoType = e.target.value;
    this.setState(() => ({ promoType }));
  }

  onDatesChange({ startDate, endDate }) {
    this.setState(() => ({ startDate, endDate }));
  }

  onFocusChange(calendarFocus) {
    this.setState(() => ({ calendarFocus }));
  }

  onSubmitReleasePage2(e) {
    e.preventDefault();
    const { error, errorsExist } = releasePage2Validation(this.state);
    this.setState(() => ({ error }));
    if (errorsExist) {
      return;
    }
    console.log(this.state);
    const currentPage = 'ReleasePage3';
    this.setState(() => ({ currentPage }));
  }

  // ReleasePage3

  onWebsiteChange(e) {
    const website = e.target.value;
    this.setState(() => ({ website }));
  }

  onKeywordsChange(e) {
    const keywords = e.target.value;
    this.setState(() => ({ keywords }));
  }

  onReleaseTextChange(e) {
    const releaseText = e.target.value;
    this.setState(() => ({ releaseText }));
  }

  onSubmitReleasePage3(e) {
    e.preventDefault();
    const { error, errorsExist } = releasePage3Validation(this.state);
    this.setState(() => ({ error }));
    if (errorsExist) {
      return;
    }
    console.log(this.state);
    const currentPage = 'SubmissionSuccess';
    this.setState(() => ({ currentPage }));
    sendReleaseToNodemailer(this.state);
  }

  onBack(e) {
    if (this.state.currentPage === 'ReleasePage2') {
      const currentPage = 'ReleasePage1';
      this.setState(() => ({ currentPage }));
    }
    if (this.state.currentPage === 'ReleasePage3') {
      const currentPage = 'ReleasePage2';
      this.setState(() => ({ currentPage }));
    }
  }

  render() {
    const {
      currentPage,
      error,
      amazonUrl,
      nonFictionOrFiction,
      genre,
      subGenre,
      email,
      price,
      promoType,
      startDate,
      endDate,
      calendarFocus,
      website,
      keywords,
      releaseText
    } = this.state;
    return (
      <ReleaseComponent
        // State
        currentPage={currentPage}
        error={error}
        // ReleasePage1
        amazonUrl={amazonUrl}
        nonFictionOrFiction={nonFictionOrFiction}
        genre={genre}
        subGenre={subGenre}
        email={email}
        // ReleasePage2
        price={price}
        promoType={promoType}
        startDate={startDate}
        endDate={endDate}
        calendarFocus={calendarFocus}
        // ReleasePage3
        website={website}
        keywords={keywords}
        releaseText={releaseText}
        // Methods
        onBack={this.onBack}
        // ReleasePage1
        onAmazonUrlChange={this.onAmazonUrlChange}
        onNonFictionOrFictionChange={this.onNonFictionOrFictionChange}
        onGenreChange={this.onGenreChange}
        onSubGenreChange={this.onSubGenreChange}
        onEmailChange={this.onEmailChange}
        onSubmitReleasePage1={this.onSubmitReleasePage1}
        // ReleasePage2
        onPriceChange={this.onPriceChange}
        onPromoTypeChange={this.onPromoTypeChange}
        onDatesChange={this.onDatesChange}
        onFocusChange={this.onFocusChange}
        onSubmitReleasePage2={this.onSubmitReleasePage2}
        // ReleasePage3
        onWebsiteChange={this.onWebsiteChange}
        onKeywordsChange={this.onKeywordsChange}
        onReleaseTextChange={this.onReleaseTextChange}
        onSubmitReleasePage3={this.onSubmitReleasePage3}
      />
    );
  }
}
