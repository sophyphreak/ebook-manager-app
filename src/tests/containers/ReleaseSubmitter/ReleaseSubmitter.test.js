import React from 'react';
import { shallow } from 'enzyme';
import ReleaseSubmitter from '../../../containers/ReleaseSubmitter/ReleaseSubmitter';

let wrapper;

beforeEach(() => {
  Date.now = jest.fn(() => 1487076708000);
  wrapper = shallow(<ReleaseSubmitter />);
});

test('should render ReleaseSubmitter correctly', () => {
  expect(wrapper).toMatchSnapshot();
});
