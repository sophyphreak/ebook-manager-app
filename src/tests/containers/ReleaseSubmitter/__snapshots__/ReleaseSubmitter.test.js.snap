// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReleaseSubmitter correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReleaseSubmitter />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "amazonUrl": "",
      "calendarFocus": null,
      "currentPage": "ReleasePage1",
      "email": "",
      "endDate": null,
      "error": Object {
        "amazonUrl": "",
        "email": "",
        "genre": "",
        "keywords": "",
        "message": "",
        "nonFictionOrFiction": "",
        "price": "",
        "promoType": "",
        "releaseText": "",
        "website": "",
      },
      "genre": "Please select",
      "keywords": "",
      "nonFictionOrFiction": "Non-fiction",
      "onAmazonUrlChange": [Function],
      "onBack": [Function],
      "onDatesChange": [Function],
      "onEmailChange": [Function],
      "onFocusChange": [Function],
      "onGenreChange": [Function],
      "onKeywordsChange": [Function],
      "onNonFictionOrFictionChange": [Function],
      "onPriceChange": [Function],
      "onPromoTypeChange": [Function],
      "onReleaseTextChange": [Function],
      "onSubGenreChange": [Function],
      "onSubmitReleasePage1": [Function],
      "onSubmitReleasePage2": [Function],
      "onSubmitReleasePage3": [Function],
      "onWebsiteChange": [Function],
      "price": "",
      "promoType": "Free (Limited Time)",
      "releaseText": "",
      "startDate": null,
      "subGenre": "",
      "website": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "amazonUrl": "",
        "calendarFocus": null,
        "currentPage": "ReleasePage1",
        "email": "",
        "endDate": null,
        "error": Object {
          "amazonUrl": "",
          "email": "",
          "genre": "",
          "keywords": "",
          "message": "",
          "nonFictionOrFiction": "",
          "price": "",
          "promoType": "",
          "releaseText": "",
          "website": "",
        },
        "genre": "Please select",
        "keywords": "",
        "nonFictionOrFiction": "Non-fiction",
        "onAmazonUrlChange": [Function],
        "onBack": [Function],
        "onDatesChange": [Function],
        "onEmailChange": [Function],
        "onFocusChange": [Function],
        "onGenreChange": [Function],
        "onKeywordsChange": [Function],
        "onNonFictionOrFictionChange": [Function],
        "onPriceChange": [Function],
        "onPromoTypeChange": [Function],
        "onReleaseTextChange": [Function],
        "onSubGenreChange": [Function],
        "onSubmitReleasePage1": [Function],
        "onSubmitReleasePage2": [Function],
        "onSubmitReleasePage3": [Function],
        "onWebsiteChange": [Function],
        "price": "",
        "promoType": "Free (Limited Time)",
        "releaseText": "",
        "startDate": null,
        "subGenre": "",
        "website": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
